src/sympc/protocol/falcon/falcon.py

```diff
@@ -63,3 +65,87 @@ def __eq__(self, other: Any) -> bool:
             return False
 
         return True
+
+    @staticmethod
+    def mul_master(x: MPCTensor, y: MPCTensor, session: Session) -> MPCTensor:
+        """Master method for multiplication.
+
+        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
+        This needs to be improved in future, it relies on orchestrator being a trusted third party.
+        Falcon, requires parties to be able to communication between each other.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+
+        Returns:
+            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+
+        Raises:
+            ValueError: Raised when number of parties are not three.
+            NotImplementedError: Raised when implementation not present
+
+        """
+        if len(session.parties) != 3:
+            raise ValueError("Falcon requires 3 parties")
+
+        result = None
+
+        if session.protocol.security_type == "semi-honest":
+            result = Falcon.multiplication_protocol(x, y)
+
+        elif session.protocol.security_type == "malicious":
+            raise NotImplementedError(
+                "Multiplication has not been implemented for malicious setting"
+            )
+        else:
+            raise NotImplementedError(
+                f"mult operation not implemented for {session.protocol.security_type} setting"
+            )
+
+        przs_masks = [
+            session.session_ptrs[index]
+            .przs_generate_random_share(shape=x.shape)
+            .get_shares()
+            .get()[0]
+            for index in range(0, 3)
+        ]
+
+        # Add random mask to vals
+        shares = [
+            result[0].get() + przs_masks[0],
+            result[1].get() + przs_masks[1],
+            result[2].get() + przs_masks[2],
+        ]
+
+        shares = ReplicatedSharedTensor.distribute_shares(shares, x.session)
+        result = MPCTensor(shares=shares, session=x.session)
+        return result
+
+    @staticmethod
+    def multiplication_protocol(
+        x: MPCTensor, y: MPCTensor
+    ) -> List[ReplicatedSharedTensor]:
+        """Implementation of Falcon's multiplication with semi-honest security guarantee.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+
+        Returns:
+            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+        """
+        z_shares = []
+
+        for party in range(0, len(x.share_ptrs)):
+
+            z_value = (
```

Comment:
Q: can't you use ```parallel_execution``` for this (also for the PRZS) from the caller. In that case you won't have that much communcation